server/src/main/java/org/opensearch/index/store/Store.java

```diff
@@ -1961,6 +1949,11 @@ public void incRefFileDeleter(Collection<String> files) {
     public void decRefFileDeleter(Collection<String> files) {
         if (this.indexSettings.isSegRepEnabled()) {
             this.replicaFileTracker.decRef(files);
+            try {
+                cleanUnReferencedReaderManagerFiles();
```

Comment:
Rather than calling this here I think we could pass the deleter a function that locks and deletes a single file when the ref count hits 0. This would prevent looping through the entire dir every time.

Reply:
Thanks @mch2 for bringing this up. I see this is an improvement which prevents going through all the files currently referenced by ReplicaFileTracker and delete when a file's reference count goes down to `0`. 

Though with above, we need to obtain `write.lock` for each file individually. At any point, replica will have limited number of files (two commits + more files preserved from pit/scroll context), there will be a limited (but unbounded) number of files referenced by ReplicaFileTracker. I am not super sure, how costly is the [filelock](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/channels/FileChannel.html#lock(long,long,boolean)) operation, provided it needs to make locking visible to processes outside of JVM.